<!DOCTYPE html>
<html>
<head>
<title>Extra_maftools_exercises.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="extra-maftools-exercises">Extra maftools exercises</h1>
<p>Maftools has other interesting options that can help you compare and analyze your data.</p>
<p>For some of this exercises we will be using some of the maf files from the TCGA that are included inside the maftools package.</p>
<h2 id="comparing-the-mutational-load-to-tcga-data">Comparing the mutational load to TCGA data</h2>
<p>As mentioned above, maftools has access to the data of 33 TCGA cohorts. We can see what cohorts are available as follows:</p>
<pre class="hljs"><code><div>tcga_data = tcgaAvailable()
<span class="hljs-comment">#This will store the information of the study names and abbreviations in "tcga_data"</span>
tcga_data
<span class="hljs-comment"># This will print the information stored </span>
</div></code></pre>
<p>We can use this data to compare the mutational load of our data to the TCGA cohorts.</p>
<p>To do so we will use the following commmand:</p>
<pre class="hljs"><code><div>BLCA.mutload = tcgaCompare(maf=BLCA, cohortName = <span class="hljs-string">'our_BLCA'</span>, logscale =TRUEpl)
</div></code></pre>
<p>As we are also working with TCGA data, you can see that our cohort is almost the same as the one included in the maftools package.</p>
<p>We can also compare the genes we have identified with dndscv to other cohort (if they are also altered or not).</p>
<pre class="hljs"><code><div>tcgaDriverBP(BLCA, genes = c(<span class="hljs-string">"TP53"</span>,<span class="hljs-string">"PIK3CA"</span>,<span class="hljs-string">"ARID1A"</span>, <span class="hljs-string">"KMT2D"</span>, <span class="hljs-string">"RB1"</span>))

</div></code></pre>
<h2 id="plotting-variant-allele-frequencies">Plotting variant allele frequencies</h2>
<p>Variant allele frequency can help us visualize clonality in our samples (But remember that other factors can affect tha VAF, such as the quality and purity of our samples). Maftools allows us to generate boxplots of this VAF. In order to do so, the package can extract the data from the maf file or if it is not stated directly in the maf, maftools can calculate it using the information of the number of reads supporting the variant and reference alleles.</p>
<p>To generate the boxplot type:</p>
<pre class="hljs"><code><div>plotVaf(maf=BLCA)
</div></code></pre>
<p>TO establish the genes we want we follow the same pattern as with other commands</p>
<pre class="hljs"><code><div>plotVaf(maf=BLCA, genes=c(<span class="hljs-string">"TP53"</span>,<span class="hljs-string">"PIK3CA"</span>,<span class="hljs-string">"ARID1A"</span>, <span class="hljs-string">"KMT2D"</span>, <span class="hljs-string">"RB1"</span>) )

</div></code></pre>
<h2 id="adding-clinical-annotations">Adding clinical annotations</h2>
<p>Oncoplots can include several annotations, usually referring to clinical data. We will now be using data from the TCGA included in maftools to practice how to add annotations on the oncoplot.</p>
<pre class="hljs"><code><div>SKCM = tcgaLoad(study=<span class="hljs-string">"SKCM"</span>)
<span class="hljs-comment">#Loading the data for skin cutaneous melanoma</span>

getClinicalData(SKCM)
<span class="hljs-comment">#This will show us the clinical data available for this cohort</span>

oncoplot(SKCM, dra_titv=<span class="hljs-literal">TRUE</span>, clinicalFeatures =<span class="hljs-string">'sample_type_description'</span>)
<span class="hljs-comment">#Here we are choosing the sampl type(metastasic or primary) to add as annotation. To show this info as annotations just add the column names of the available clinical data for this cohort. </span>

</div></code></pre>
<p>Maftools still has a lot of other options that can be useful, from things like extracting data from maf files, to othe type of visualizations and annotations with copy number data and mutational signatures. All the options available are found in their user manual.</p>
<p>As well, remember there are other options to generate oncoplots and other types of visualization of your data, many of them as R packages. An example is ComplexHeatmap, which also has an option to generate an oncoplot and has interactive options.</p>

</body>
</html>
